grpc/src/main/java/com/linecorp/armeria/server/grpc/DeferredListener.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2023 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.grpc;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+
+final class DeferredListener<I> extends ServerCall.Listener<I> {
+
+    private static final List<?> NOOP_TASKS = ImmutableList.of();
+
+    private List<Consumer<Listener<I>>> pendingTasks = new ArrayList<>();
+    @Nullable
+    private Listener<I> delegate;
+    private boolean callClosed;
```

Comment:
Good point. I didn't think of that case. 😅
`volatile` keyword wouldn't be enough because we need to consider both `delegate` and `pendingTasks. Let me try to use a lock to make it thread-safe.

Reply:
Checked the Javadoc of `ServerCall.Listener` and it is not allowed to concurrently access `ServerCall.Listener` by two threads.
https://github.com/grpc/grpc-java/blob/fcb5c54e4b82d354f42ced0121928fabce9ef53f/api/src/main/java/io/grpc/ServerCall.java#L50
It would be better to guarantee the access order using `serialExecutor`.